src/endpoint/s3/s3_rest.js

```diff
@@ -229,16 +229,22 @@ async function authorize_request_policy(req) {
     }
 
     const account = req.object_sdk.requesting_account;
-    const account_identifier = req.object_sdk.nsfs_config_root ? account.name.unwrap() : account.email.unwrap();
-    const is_system_owner = account_identifier === system_owner.unwrap();
+    const account_identifier_name = req.object_sdk.nsfs_config_root ? account.name.unwrap() : account.email.unwrap();
+    const account_identifier_id = req.object_sdk.nsfs_config_root ? account._id : undefined;
+    //In old buckets (until 5.17), system_owner is account name (sensitive string).
+    //In new buckets (since 5.18), system_owner is account id (string).
+    const system_owner_equatable = system_owner.unwrap ? system_owner.unwrap() : system_owner;
+    const is_system_owner =
+        (account_identifier_name === system_owner_equatable) ||
+        (account_identifier_id && account_identifier_id === system_owner_equatable);
 
     // @TODO: System owner as a construct should be removed - Temporary
     if (is_system_owner) return;
 
     const is_owner = (function() {
         if (account.bucket_claim_owner && account.bucket_claim_owner.unwrap() === req.params.bucket) return true;
-        if (owner_account && owner_account.id === account._id) return true;
-        if (account_identifier === bucket_owner.unwrap()) return true;
+        if (req.object_sdk.nsfs_config_root && account._id === owner_account.id) return true; // NC NSFS case
+        if (!req.object_sdk.nsfs_config_root && account_identifier_name === bucket_owner.unwrap()) return true;
```

Comment:
I don't think that s3 rest api should look at nsfs_config_root at all, this is why we have an AccountSpace and BucketSpace interfaces. It seems that the use of this variable somehow has spread around in the code (even into ssl_utils god forbid). Instead, the code that does not directly writes/reads from the config dir should just look at the actual bucket and account config structures it received from read_bucket_sdk_policy_info and requesting_account, and use their information to make the ownership checks.

Reply:
Generally I agree, but practically I went with the existing-
const account_identifier = req.object_sdk.nsfs_config_root ? account.name.unwrap() : account.email.unwrap();

I can try to remove what I have added.

Reply:
I looked at it, and I don't see a quick/simple solution to differentiate between by-name or by-id ownership (in s3_rest).
If there's something I'm missing please let me know.
I can go by whether account_identifier_id is defined to reduce the foot print of req.object_sdk.nsfs_config_root in the code (but it's equivalent).